realhf/impl/model/modules/mlp.py

```diff
@@ -306,42 +307,52 @@ def __init__(
         hidden_dim: int,
         intermediate_dim: int,
         activation_function: str,
-        layer_norm_epsilon: float,
-        layer_norm_type: str,
+        # layer norm
+        layer_norm_epsilon: float = 1e-5,
+        layer_norm_type: str = "rms",
         # parallelism
         model_parallel: bool = False,  # We set this as an option for replacing this module with layers in transformer engine
         gradient_accumulation_fusion: bool = False,
+        # whether this MLP is used as expert
+        is_expert: bool = False,
         # dtype and device
         dtype: Optional[torch.dtype] = None,
         device: Optional[Union[str, torch.device]] = None,
     ):
         super().__init__()
-        sequence_parallel = constants.sequence_parallel()
-        if not model_parallel and (sequence_parallel or gradient_accumulation_fusion):
+        # when used as experts the MLP always compute without sequence parallel
+        self.sequence_parallel = constants.sequence_parallel() and not is_expert
```

Comment:
Why not combine sequence parallel and MoE?

Reply:
This is because the gather and scatter of sequence parallel are done in MoE token dispatcher before and after permutation. I did not re-scatter them again before feeding permuted hidden states into experts because it seems meaningless and only add communication overheads.  Therefore even when sequence parallel is enabled, expert MLPs behave as sequence parallel is not enabled. 

Maybe there is a clever way to permute hidden states with sequence parallel to save memory and reduce communication overhead but  I did not think of one.